src/ray/common/asio/periodical_runner.cc

```diff
@@ -38,103 +33,106 @@ void PeriodicalRunner::Clear() {
-          // gcs_server)
-          return;
+      [weak_self = weak_from_this(),
+       fn = std::move(fn),
+       period,
+       timer = std::move(timer)](const boost::system::error_code &error) mutable {
+        if (auto self = weak_self.lock(); self) {
+          if (error == boost::asio::error::operation_aborted) {
+            // `operation_aborted` is set when `timer` is canceled or destroyed.
+            // The Monitor lifetime may be short than the object who use it. (e.g.
+            // gcs_server)
+            return;
+          }
+          RAY_CHECK(!error) << error.message();
+          self->DoRunFnPeriodically(std::move(fn), period, std::move(timer));
         }
-        RAY_CHECK(!error) << error.message();
-        DoRunFnPeriodically(fn, period, timer);
       });
 }
 
 void PeriodicalRunner::DoRunFnPeriodicallyInstrumented(
-    const std::function<void()> &fn,
+    std::function<void()> fn,
     boost::posix_time::milliseconds period,
     std::shared_ptr<boost::asio::deadline_timer> timer,
-    const std::string name) {
+    std::string name) {
   fn();
   absl::MutexLock lock(&mutex_);
   timer->expires_from_now(period);
   // NOTE: We add the timer period to the enqueue time in order only measure the time in
   // which the handler was elgible to execute on the event loop but was queued by the
   // event loop.
   auto stats_handle = io_service_.stats().RecordStart(name, period.total_nanoseconds());
-  timer->async_wait([this,
-                     fn = std::move(fn),
-                     stopped = stopped_,
-                     period,
-                     timer = std::move(timer),
-                     stats_handle = std::move(stats_handle),
-                     name](const boost::system::error_code &error) {
-    if (*stopped) {
-      return;
-    }
-    io_service_.stats().RecordExecution(
-        [this,
-         stopped = stopped_,
-         fn = std::move(fn),
-         error,
-         period,
-         timer = std::move(timer),
-         name]() {
-          if (*stopped) {
-            return;
-          }
-          if (error == boost::asio::error::operation_aborted) {
-            // `operation_aborted` is set when `timer` is canceled or destroyed.
-            // The Monitor lifetime may be short than the object who use it. (e.g.
-            // gcs_server)
-            return;
-          }
-          RAY_CHECK(!error) << error.message();
-          DoRunFnPeriodicallyInstrumented(fn, period, timer, name);
-        },
-        std::move(stats_handle));
-  });
+  timer->async_wait(
+      [weak_self = weak_from_this(),
+       fn = std::move(fn),
+       period,
+       timer = std::move(timer),
+       stats_handle = std::move(stats_handle),
+       name = std::move(name)](const boost::system::error_code &error) mutable {
+        if (auto self = weak_self.lock(); self) {
+          self->io_service_.stats().RecordExecution(
+              [weak_self,
```

Comment:
we can just capture `self` instead of `weak_self` since `RecordExecution` runs the lambda immediately.

Reply:
done